File: image_processing.py
from PIL import Image  # pip install pillow
import numpy as np
import scipy.ndimage as nd
from scipy.misc import imsave


#def rgb2gray(rgb):
#
#    r, g, b = rgb[:,:,0], rgb[:,:,1], rgb[:,:,2]
#    gray = 0.2989 * r + 0.5870 * g + 0.1140 * b
#
#    return gray


def rgb_to_gray(rgb_image):
    return (np.dot(rgb_image[...,:3], [0.299, 0.587, 0.114])) / 255.0


def resize_image(image, new_shape):
    '''resize image to have specified shape new_shape
    image: np.array (2D or 3D)
    new_shape: tuple with number of pixels in width and height of new image
    '''

    # numpy to PIL
    image_pil = Image.fromarray(image)

    image_pil = image_pil.resize(new_shape, Image.BILINEAR)

    # PIL to numpy
    image = np.asarray(image_pil)

    return image


def red_to_white(image):
    '''image: np.array with dimension m*n*3 (image with RGB layers)
    converts all red pixels to white'''

    red_pixels = ((image[:,:,0] > 200) & (image[:,:,1] < 20) & (image[:,:,2] < 20))
    image[red_pixels,:] = 255

    return image


def find_crop_index(image, axis):
    '''along the dimension axis, find the index of the first non-white pixel
    input:
        image: np.array with dimension m*n containing floats between 0 and 1 (grey  scale image)
        axis: np array axis (0=rows, 1=columns)
    '''
    white_cut = 1.0  # value below which we consider pixels to be non-white

    # get the first index in each col/row along that axis where there is a non-white pixel
    first_non_white = np.apply_along_axis(np.argmax, axis, image < white_cut)

    # get the smallest non-zero value
    # (np.argmax returns 0 when no index is found)
    index = np.min(np.extract(first_non_white > 0, first_non_white))

    return index


def flip(image, axis):
    '''reimplement flip function from numpy v1.12.0'''
    if axis == 0:
        return np.flipud(image)
    else:
        return np.fliplr(image)


def crop_whitespace(image):
    '''crop whitespace around image, up until the first non-white
    pixel found in each direction
    image: np.array with dimension m*n containing floats between 0 and 1 (greyscale image)'''

    for axis in [0, 1]:
        # get bounding box for this axis
        min_index = find_crop_index(image, axis)
        max_index = image.shape[axis] - find_crop_index(flip(image, axis), axis)
        # crop this axis
        image = np.take(image, range(min_index, max_index), axis=axis)

    return image

File: webapp/js/image_processing.js
// get the white pixels from the black canvas
// crop image
// resize
// convert to black on white (values between 0 and 1)
// return array of floats between 0 and 1 of length modelImageSize*modelImageSize
function extractData() {

    // first extract image data from the visible canvas
    var imgData = ctx.getImageData(0, 0, canvas.width, canvas.height);

    // get definition of crop box (smallest box bounding all white pixels)
    var cropBox = cropWhitespace(imgData);
    console.log(cropBox);

    // resize data in crop box to be modelImageSize*modelImageSize
    var croppedRescaledImgData = resizeImage(cropBox.cropX, cropBox.cropY, cropBox.cropWidth, cropBox.cropHeight, modelImageSize, modelImageSize);

    // image is black and white, all channels the same
    // therefore we simply take the pixels of the first channel (R)
    var data = new Array();
    for (var i = 0; i < croppedRescaledImgData.data.length; i += 4) {
        data.push(croppedRescaledImgData.data[i]);
    }
    console.log(data);

    // convert to black on white, values scaled between 0 and 1
    // data = convert();

    return data;
}

// along the dimension 0, find the indices of the first and last white pixels
// input:
//     rows: array of arrays containing integers between 0 and 255
function findCropIndices(rows) {

    var firstWhiteIndex = [];
    var lastWhiteIndex = [];

    // find first and last white pixels in each row
    for (var ir = 0; ir < rows.length; ir += 1) {
        var ifirst = rows[ir].findIndex(r => r == 255);
        if (ifirst > 0) { firstWhiteIndex.push(ifirst); }
        var ilast = rows[ir].reverse().findIndex(r => r == 255);
        if (ilast > 0) { lastWhiteIndex.push(rows[ir].length - ilast); }
    }

    // find min and max over all rows
    // TODO deal with case where no white indices found
    var minI = Math.min.apply(Math, firstWhiteIndex);
    var maxI = Math.max.apply(Math, lastWhiteIndex);

    return {'minWhitePixel': minI, 'maxWhitePixel': maxI};
}

// crop whitespace around symbol, up until the first non-white pixel found in each direction
// imgData: ImageData object
// symbol is drawn in white on black, all channels the same
// and lines dividing the three zones are green
// therefore we simply work with the pixels of the red channel 
function cropWhitespace(imgData) {

    var data = imgData.data;
    var nPointsPerRow = imgData.width * 4 // 4 channels
    var nPointsPerCol = imgData.height * 4 // 4 channels

    // get array of rows and cols in R channel
    var rows = [];
    var cols = [];
    for (var j = 0; j < nPointsPerRow; j += 4) {  // red channel only
        var col = [];
        cols.push(col);
    }
    for (var i = 0; i < data.length; i += nPointsPerRow) {  //  
        var row = [];
        for (var j = i; j < (i + nPointsPerRow); j += 4) {  // red channel only
            row.push(data[j]);
            var ic = (j - i) / 4;
            cols[ic].push(data[j]);
        }
        rows.push(row);
    } 

    // get first and last white pixels in each direction
    var rowsIndices = findCropIndices(rows);
    var colsIndices = findCropIndices(cols);
    console.log(rowsIndices);
    console.log(colsIndices);

    // get bounding box in each direction
    var cropX = rowsIndices.minWhitePixel;
    var cropY = colsIndices.minWhitePixel;
    var cropWidth = rowsIndices.maxWhitePixel - rowsIndices.minWhitePixel;
    var cropHeight = colsIndices.maxWhitePixel - colsIndices.minWhitePixel;

    return {'cropX': cropX, 'cropY': cropY, 'cropWidth': cropWidth, 'cropHeight': cropHeight};
}

// Take the sub-rectangle defined by sourceX, sourceY, sourceWidth, sourceHeight
// of the image currently drawn on the global canvas,
// and output that image array rescaled to newWidth and newHeight.
// Function rescales image by writing to an invisible canvas with context resizingCtx
// and returns the cropped, rescaled imgData extracted from that invisible canvas.
function resizeImage(sourceX, sourceY, sourceWidth, sourceHeight, newWidth, newHeight) {
    ctx.drawImage(canvas, sourceX, sourceY, 
    //resizingCtx.drawImage(sourceX, sourceY, 
                          sourceWidth, sourceHeight, 
                          0, 0, newWidth, newHeight);
    var imgData = resizingCtx.getImageData(0, 0, newWidth, newHeight);
    console.log(imgData);
    return imgData;
}
